### ISSUE TITLE
Use global timeout if global timeout > local timeout

### ISSUE THREAD

**Issue by hverlin** [mochajs/mocha#4536]
**Is your feature request related to a problem or a nice-to-have? Please describe.**
When running mocha locally, you might want to sometimes raise the timeout for a test if you know that it will be slow with e.g.
```
this.timeout(3000)
```
so that it is greater than the value set globally (with `-t`, default being `2000`). 

The issue comes when running in tests in parallel where all tests are usually running slower individually.
We will e.g. set the global timeout to 10 000ms to take the parallelism into account.

The problem is that the value of `this.timeout(3000)` will take priority for this test.

Consider the following test that would run fine sequentially, but would maybe fail when run in parallel
```
describe('a test', () => {
  it('should run fine in parallel', function(done) {
    this.timeout(110);
    // adding timeout so that the test fails
    setTimeout(done, 100);
  });
});
```

**Describe the solution you'd like**
I would like `mocha` to use the global timeout if global timeout > local timeout.
Alternatively, I would like to express the test timeout as a function of the global timeout.

The `timeout` function should only be used to detect tests that might be stuck, so it should IMO be fine to make this breaking change. 

See https://developer.ibm.com/languages/node-js/articles/parallel-tests-mocha-v8/ (from @boneskull)
> Mocha’s “timeout” functionality is not to be used as a benchmark; its intent is to catch code that takes an unexpectedly long time to execute. Since we now expect tests to potentially take longer, we can safely increase the timeout value.

Alternatively, allow passing an argument to ignore the local timeout if the global timeout is larger.

**Describe alternatives you've considered**
Right now, we need to either:
- make sure developers don't use `timeout` which would be too low to run in parallel
- don't use `this.timeout(xxx)` but our own function that would check for any global timeout
- override the `timeout` functions, e.g. `mocha.Mocha.Suite.prototype.timeout`
- Fork mocha to add this change

**Additional context**
I expect that these kinds of issues will appear a lot more since mocha now introduced running tests in parallel.

**Comment by outsideris** [comment]
Setting a timeout for a specific function means that the developer has a reason for that test to have a specific timeout, so I don't think the global timeout should take precedence.

A global timeout is a timeout that applies to the entire test, and I think it is a more common use case for developers to use `this.timeout()` when they want a specific test to have a different timeout.

**Comment by hverlin** [comment]
Since the timeout feature is not supposed to be used as a benchmark, I don't see the reason for not taking the max of (localTimeout, globalTimeout).

Can you describe me a use case where this would make sense? 

When running tests in parallel, the global timeout should take precedence if it is higher IMO since the author of the test would not know about the amount of parallelism the CI server will use.

**Comment by outsideris** [comment]
I think I didn't fully understand what you said. 

Why the following test will be failed in parallel mode? 
```
describe('a test', () => {
  it('should run fine in parallel', function(done) {
    this.timeout(110);
    // adding timeout so that the test fails
    setTimeout(done, 100);
  });
});
```

**Comment by hverlin** [comment]
it is explained in the link I have mentioned: https://developer.ibm.com/languages/node-js/articles/parallel-tests-mocha-v8/ 
> Because of many variables — from Mocha to Node.js to the OS to the CPU itself — parallel mode exhibits a much wider range of timings for any given test. These timeout exceptions don’t indicate a newfound performance issue; rather, they’re a symptom of a naturally higher system load and nondeterministic execution order.

![image](https://user-images.githubusercontent.com/9151470/104009902-054ba380-51ac-11eb-94e9-506ea0a2880b.png)

**Comment by juergba** [comment]
> I would like mocha to use the global timeout if global timeout > local timeout.

I don't like this idea, because the rule is just not logical. In Javascript a variable `timeout` of the inner scope shadows `timeout` of the outer scope. Any different behaviour causes confusion, and not many users actually read our documentation.

You could use `this.timeout(timeFactor * 110)` where ` timeFactor` depends on your testing environment and defaults to 1. You could set this factor via an environment variable.
